pinot-segment-local/src/main/java/org/apache/pinot/segment/local/data/manager/TableDataManager.java

```diff
@@ -78,6 +80,14 @@ void addSegment(File indexDir, IndexLoadingConfig indexLoadingConfig)
   void addSegment(String segmentName, TableConfig tableConfig, IndexLoadingConfig indexLoadingConfig)
       throws Exception;
 
+  /**
+   * Add or replace an immutable segment for current table, which can be either
+   * OFFLINE or REALTIME table.
+   */
+  void addOrReplaceSegment(String segmentName, IndexLoadingConfig indexLoadingConfig,
+      @Nullable SegmentMetadata localMetadata, SegmentZKMetadata zkMetadata, boolean forceDownload)
```

Comment:
(nit) for slightly better readability:
```suggestion
      SegmentZKMetadata remoteZKMetadata, @Nullable SegmentMetadata localMetadata, boolean forceDownload)
```